tracer/src/Datadog.Trace/Debugger/Instrumentation/AsyncMethodDebuggerInvoker.cs

```diff
@@ -0,0 +1,367 @@
+        // For case 1 it's simple because we can store all data in AsyncLocal and for each entry to the BeginMethod (through the state machine object)
+        // we can inspect the AsyncLocal to see if we already create the context ot not
+        // but for case 2 and 2.1 the AsyncLocal of the callee will drive the context from the caller
+        // so we can't deterministic know if the context is a continuation or first entry of the callee
+
+        // We use the AsyncLocal storage to store the debugger state of each logical invocation of the async method i.e. StateMachine.MoveNext run
+        // We want to have the same debugger state when reentry in same context (e.g. continuation after executing an "await" expression) but a different debugger state when the entry is in a different context
+
+        /// <summary>
+        /// We use the AsyncLocal storage to store the debugger state of each logical invocation of the async method i.e. StateMachine.MoveNext run
+        /// We want to have the same debugger state when reentry in same context (e.g. continuation after executing an "await" expression) but a different debugger state when the entry is in a different context
+        /// </summary>
+        private static readonly AsyncLocal<AsyncMethodDebuggerState> AsyncContext = new();
+
+        /// <summary>
+        /// Create a new context and save the parent.
+        /// We call this method from the async kick off method before the state machine is created,
+        /// so for each async state machine, this method will be called exactly once.
+        /// The IsFirstEntry property of the AsyncMethodDebuggerState will be "true"
+        /// then, in the first state machine step, we set this property to "false", thus
+        /// we can deduce that any further invocation is not a new logical call but rather a
+        /// continuation from an "await" expression.
+        /// </summary>
+        /// <param name="kickoffInfo">The async kickoff method info</param>
+        /// <returns>Live debugger state</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static AsyncMethodDebuggerState SetContext(AsyncHelper.AsyncKickoffMethodInfo kickoffInfo)
+        {
+            var currentState = AsyncContext.Value;
+            var newState = new AsyncMethodDebuggerState
+            {
+                Parent = currentState,
+                KickoffInvocationTarget = kickoffInfo.KickoffParentObject,
+                KickoffInvocationTargetType = kickoffInfo.KickoffParentType,
+                KickoffMethod = kickoffInfo.KickoffMethod
+            };
+            AsyncContext.Value = newState;
+            return newState;
+        }
+
+        /// <summary>
+        /// Restore the async context to the parent
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void RestoreContext()
```

Comment:
What happens if an exception was thrown from inside our instrumentation and then we mark the state as 'invalid', will we restore the context then? it might break the whole tree of async locals.
We need to make sure we restore the context correctly, regardless of exceptions that might occur during the execution. (Maybe the fix will be to just add a call to `RestoreContext` from inside the `LogException`, but it needs to be verified)

Reply:
Fixed in 871a652d9b150d63443d87a542e34d4643145adf